File: programming_methods/laboratories/5/5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>

size_t n, m;
const long long INF = LONG_LONG_MAX;

struct Edge {
    int from, to;
    long long flow, cap;
    int reversed = -1;

    Edge(int from, int to, long long f, long long cap) : from(from), to(to), flow(f), cap(cap) {}
};

std::vector<std::vector<Edge>> edges;
std::vector<long long> d;
std::vector<int> id;
std::vector<Edge*> p;

void addEdge(int from, int to, long long flow, long long cap) {
    Edge edge = Edge(from, to, flow, cap);
    Edge rev = Edge(to, from, flow, 0);

    edges[from].push_back(edge);
    edges[to].push_back(rev);

    edges[from].back().reversed = edges[to].size() - 1;
    edges[to].back().reversed = edges[from].size() - 1;
}

long long dinicAlgorithm() {
    long long maxFlow = 0;
    while (true) {
        // Очереди для BFS
        std::queue<int> q1, q2;
        id.assign(n + 1, 0);
        d.assign(n + 1, INF);
        p.resize(n + 1);

        // Инициализация BFS
        d[1] = 0;
        q1.push(1);
        id[1] = 1;

        while (!q1.empty() || !q2.empty()) {
            // Обрабатываем первую очередь
            if (!q1.empty()) {
                int u = q1.front();
                q1.pop();
                id[u] = 2;

                for (Edge& edge : edges[u]) {
                    if (edge.flow < edge.cap && d[edge.to] > d[edge.from] + 1) {
                        d[edge.to] = d[edge.from] + 1;
                        if (id[edge.to] == 0) {
                            q2.push(edge.to); // Переводим в очередь q2 для дальнейшей обработки
                            id[edge.to] = 1;
                            p[edge.to] = &edge;
                        }
                    }
                }
            }

            // Обрабатываем вторую очередь
            if (!q2.empty()) {
                int u = q2.front();
                q2.pop();
                id[u] = 2;

                for (Edge& edge : edges[u]) {
                    if (edge.flow < edge.cap && d[edge.to] > d[edge.from] + 1) {
                        d[edge.to] = d[edge.from] + 1;
                        if (id[edge.to] == 0) {
                            q1.push(edge.to); // Переводим обратно в очередь q1
                            id[edge.to] = 1;
                            p[edge.to] = &edge;
                        }
                    }
                }
            }
        }

        long long del = INF;
        if (d[n] == INF) {
            break;
        } else {
            for (int u = n; u != 1; u = p[u]->from) {
                Edge* edge = p[u];
                del = std::min(del, edge->cap - edge->flow);
            }

            // Обновление потоков
            for (int u = n; u != 1; u = p[u]->from) {
                Edge* edge = p[u];
                Edge* reversed = &edges[edge->to][edge->reversed];

                edge->flow += del;
                reversed->flow -= del;
            }

            maxFlow += del;
        }
    }

    return maxFlow;
}

int main() {
    std::cin >> n >> m;
    edges.assign(n + 1, std::vector<Edge>());

    // Чтение рёбер
    for (int i = 0; i < m; ++i) {
        int from, to;
        long long cap;
        std::cin >> from >> to >> cap;

        addEdge(from, to, 0, cap);
    }

    // Вывод максимального потока
    std::cout << dinicAlgorithm() << std::endl;
    return 0;
}

File: programming_methods/laboratories/5/5.py
from collections import defaultdict, deque


def edmonds_karp(n, edges):
    # Создаем граф с пропускными способностями
    capacity = defaultdict(lambda: defaultdict(int))
    for u, v, c in edges:
        capacity[u][v] += c  # Учитываем возможность нескольких рёбер между вершинами

    # Инициализация переменных
    source, sink = 1, n
    flow = 0

    while True:
        # BFS для поиска увеличивающего пути
        parent = [-1] * (n + 1)
        parent[source] = source
        queue = deque([source])
        path_flow = {source: float('Inf')}

        while queue:
            current = queue.popleft()
            for next_vertex in capacity[current]:
                if parent[next_vertex] == -1 and capacity[current][
                    next_vertex] > 0:  # Остаточная пропускная способность > 0
                    parent[next_vertex] = current
                    path_flow[next_vertex] = min(path_flow[current], capacity[current][next_vertex])
                    if next_vertex == sink:
                        break
                    queue.append(next_vertex)
            else:
                continue
            break

        # Если увеличивающий путь не найден, выходим из цикла
        if parent[sink] == -1:
            break

        # Увеличиваем поток по найденному пути
        increment = path_flow[sink]
        flow += increment
        v = sink
        while v != source:
            u = parent[v]
            capacity[u][v] -= increment
            capacity[v][u] += increment
            v = u

    return flow


# Считывание входных данных
n, m = map(int, input().split())
edges = []
for _ in range(m):
    u, v, c = map(int, input().split())
    edges.append((u, v, c))

# Вычисление максимального потока
max_flow = edmonds_karp(n, edges)
print(max_flow)